File: fri3d-runtime/src/canvas.rs
use crate::font::{FontData, FontDrawTarget};
use crate::types::{Color, Font};
use crate::{SCREEN_HEIGHT, SCREEN_WIDTH};

pub struct Canvas {
    width: u32,
    height: u32,
    buffer: Vec<u8>,
    color: Color,
    font: Font,
    font_data: FontData,
}

impl Canvas {
    pub fn new() -> Self {
        let width = SCREEN_WIDTH;
        let height = SCREEN_HEIGHT;
        let font = Font::Primary;
        let font_data = FontData::from_font(font);
        Self {
            width,
            height,
            buffer: vec![0; (width * height) as usize],
            color: Color::Black,
            font,
            font_data,
        }
    }

    pub fn width(&self) -> u32 {
        self.width
    }

    pub fn height(&self) -> u32 {
        self.height
    }

    pub fn buffer(&self) -> &[u8] {
        &self.buffer
    }

    pub fn clear(&mut self) {
        self.buffer.fill(0);
    }

    pub fn set_color(&mut self, color: Color) {
        self.color = color;
    }

    pub fn set_font(&mut self, font: Font) {
        self.font = font;
        self.font_data = FontData::from_font(font);
    }

    pub fn color(&self) -> Color {
        self.color
    }

    pub fn font(&self) -> Font {
        self.font
    }

    pub fn draw_dot(&mut self, x: i32, y: i32) {
        self.set_pixel_with_color(x, y, self.color);
    }

    pub fn draw_line(&mut self, x1: i32, y1: i32, x2: i32, y2: i32) {
        let mut x0 = x1;
        let mut y0 = y1;
        let dx = (x2 - x1).abs();
        let sx = if x1 < x2 { 1 } else { -1 };
        let dy = -(y2 - y1).abs();
        let sy = if y1 < y2 { 1 } else { -1 };
        let mut err = dx + dy;

        loop {
            self.set_pixel_with_color(x0, y0, self.color);
            if x0 == x2 && y0 == y2 {
                break;
            }
            let e2 = 2 * err;
            if e2 >= dy {
                err += dy;
                x0 += sx;
            }
            if e2 <= dx {
                err += dx;
                y0 += sy;
            }
        }
    }

    pub fn draw_frame(&mut self, x: i32, y: i32, width: u32, height: u32) {
        if width == 0 || height == 0 {
            return;
        }
        let w = width as i32;
        let h = height as i32;
        self.draw_hline_color(x, y, width, self.color);
        self.draw_hline_color(x, y + h - 1, width, self.color);
        self.draw_vline_color(x, y, height, self.color);
        self.draw_vline_color(x + w - 1, y, height, self.color);
    }

    pub fn draw_box(&mut self, x: i32, y: i32, width: u32, height: u32) {
        if width == 0 || height == 0 {
            return;
        }
        for row in 0..height {
            self.draw_hline_color(x, y + row as i32, width, self.color);
        }
    }

    pub fn draw_rframe(&mut self, x: i32, y: i32, width: u32, height: u32, radius: u32) {
        if width == 0 || height == 0 {
            return;
        }
        let mut r = radius as i32;
        let w = width as i32;
        let h = height as i32;
        r = r.min(w / 2).min(h / 2);
        if r <= 0 {
            self.draw_frame(x, y, width, height);
            return;
        }

        self.draw_hline_color(x + r, y, (w - 2 * r) as u32, self.color);
        self.draw_hline_color(x + r, y + h - 1, (w - 2 * r) as u32, self.color);
        self.draw_vline_color(x, y + r, (h - 2 * r) as u32, self.color);
        self.draw_vline_color(x + w - 1, y + r, (h - 2 * r) as u32, self.color);

        self.draw_circle_quadrants(x + r, y + r, r, true, false, false, false);
        self.draw_circle_quadrants(x + w - 1 - r, y + r, r, false, true, false, false);
        self.draw_circle_quadrants(x + r, y + h - 1 - r, r, false, false, true, false);
        self.draw_circle_quadrants(x + w - 1 - r, y + h - 1 - r, r, false, false, false, true);
    }

    pub fn draw_rbox(&mut self, x: i32, y: i32, width: u32, height: u32, radius: u32) {
        if width == 0 || height == 0 {
            return;
        }
        let mut r = radius as i32;
        let w = width as i32;
        let h = height as i32;
        r = r.min(w / 2).min(h / 2);
        if r <= 0 {
            self.draw_box(x, y, width, height);
            return;
        }

        self.draw_box(x + r, y, (w - 2 * r) as u32, height);
        self.draw_box(x, y + r, r as u32, (h - 2 * r) as u32);
        self.draw_box(x + w - r, y + r, r as u32, (h - 2 * r) as u32);

        self.draw_disc(x + r, y + r, r as u32);
        self.draw_disc(x + w - 1 - r, y + r, r as u32);
        self.draw_disc(x + r, y + h - 1 - r, r as u32);
        self.draw_disc(x + w - 1 - r, y + h - 1 - r, r as u32);
    }

    pub fn draw_circle(&mut self, x: i32, y: i32, radius: u32) {
        if self.color == Color::Xor {
            self.draw_xor_circle(x, y, radius);
            return;
        }
        let r = radius as i32;
        if r <= 0 {
            self.set_pixel_with_color(x, y, self.color);
            return;
        }
        let mut dx = 0;
        let mut dy = r;
        let mut d = 1 - r;
        while dx <= dy {
            self.draw_circle_points(x, y, dx, dy, self.color);
            if d < 0 {
                d += 2 * dx + 3;
            } else {
                d += 2 * (dx - dy) + 5;
                dy -= 1;
            }
            dx += 1;
        }
    }

    pub fn draw_disc(&mut self, x0: i32, y0: i32, radius: u32) {
        if radius == 0 {
            self.set_pixel_with_color(x0, y0, self.color);
            return;
        }
        let r = radius as i32;
        let r_sq = r * r;
        for dy in -r..=r {
            let y_sq = dy * dy;
            if y_sq > r_sq {
                continue;
            }
            let mut x_extent = 0;
            while (x_extent + 1) * (x_extent + 1) <= r_sq - y_sq {
                x_extent += 1;
            }
            let line_x = x0 - x_extent;
            let line_y = y0 + dy;
            let line_w = (2 * x_extent + 1) as u32;
            self.draw_hline_color(line_x, line_y, line_w, self.color);
        }
    }

    pub fn draw_str(&mut self, x: i32, y: i32, text: &str) {
        let font_data = self.font_data.clone();
        font_data.draw_str(self, x, y, text, self.color);
    }

    pub fn string_width(&self, text: &str) -> u32 {
        self.font_data.string_width(text)
    }

    fn set_pixel_with_color(&mut self, x: i32, y: i32, color: Color) {
        if x < 0 || y < 0 {
            return;
        }
        let x = x as u32;
        let y = y as u32;
        if x >= self.width || y >= self.height {
            return;
        }
        let idx = (y * self.width + x) as usize;
        match color {
            Color::White => self.buffer[idx] = 0,
            Color::Black => self.buffer[idx] = 1,
            Color::Xor => self.buffer[idx] ^= 1,
        }
    }

    fn draw_hline_color(&mut self, x: i32, y: i32, length: u32, color: Color) {
        if length == 0 {
            return;
        }
        for offset in 0..length {
            self.set_pixel_with_color(x + offset as i32, y, color);
        }
    }

    fn draw_vline_color(&mut self, x: i32, y: i32, length: u32, color: Color) {
        if length == 0 {
            return;
        }
        for offset in 0..length {
            self.set_pixel_with_color(x, y + offset as i32, color);
        }
    }

    fn draw_circle_points(&mut self, x0: i32, y0: i32, dx: i32, dy: i32, color: Color) {
        self.set_pixel_with_color(x0 + dx, y0 + dy, color);
        self.set_pixel_with_color(x0 - dx, y0 + dy, color);
        self.set_pixel_with_color(x0 + dx, y0 - dy, color);
        self.set_pixel_with_color(x0 - dx, y0 - dy, color);
        self.set_pixel_with_color(x0 + dy, y0 + dx, color);
        self.set_pixel_with_color(x0 - dy, y0 + dx, color);
        self.set_pixel_with_color(x0 + dy, y0 - dx, color);
        self.set_pixel_with_color(x0 - dy, y0 - dx, color);
    }

    fn draw_xor_circle(&mut self, x0: i32, y0: i32, radius: u32) {
        if radius == 0 {
            self.set_pixel_with_color(x0, y0, Color::Xor);
            return;
        }

        let mut x = 0i32;
        let mut y = radius as i32;
        let mut d = 1 - y;

        while x <= y {
            if x == y {
                self.set_pixel_with_color(x0 + x, y0 + y, Color::Xor);
                self.set_pixel_with_color(x0 - x, y0 + y, Color::Xor);
                self.set_pixel_with_color(x0 + x, y0 - y, Color::Xor);
                self.set_pixel_with_color(x0 - x, y0 - y, Color::Xor);
            } else if x == 0 {
                self.set_pixel_with_color(x0, y0 + y, Color::Xor);
                self.set_pixel_with_color(x0, y0 - y, Color::Xor);
                self.set_pixel_with_color(x0 + y, y0, Color::Xor);
                self.set_pixel_with_color(x0 - y, y0, Color::Xor);
            } else {
                self.set_pixel_with_color(x0 + x, y0 + y, Color::Xor);
                self.set_pixel_with_color(x0 - x, y0 + y, Color::Xor);
                self.set_pixel_with_color(x0 + x, y0 - y, Color::Xor);
                self.set_pixel_with_color(x0 - x, y0 - y, Color::Xor);
                self.set_pixel_with_color(x0 + y, y0 + x, Color::Xor);
                self.set_pixel_with_color(x0 - y, y0 + x, Color::Xor);
                self.set_pixel_with_color(x0 + y, y0 - x, Color::Xor);
                self.set_pixel_with_color(x0 - y, y0 - x, Color::Xor);
            }

            if d < 0 {
                d += 2 * x + 3;
            } else {
                d += 2 * (x - y) + 5;
                y -= 1;
            }
            x += 1;
        }
    }

    fn draw_circle_quadrants(
        &mut self,
        x0: i32,
        y0: i32,
        radius: i32,
        top_left: bool,
        top_right: bool,
        bottom_left: bool,
        bottom_right: bool,
    ) {
        if radius <= 0 {
            return;
        }
        let mut dx = 0;
        let mut dy = radius;
        let mut d = 1 - radius;

        while dx <= dy {
            if top_left {
                self.set_pixel_with_color(x0 - dx, y0 - dy, self.color);
                self.set_pixel_with_color(x0 - dy, y0 - dx, self.color);
            }
            if top_right {
                self.set_pixel_with_color(x0 + dx, y0 - dy, self.color);
                self.set_pixel_with_color(x0 + dy, y0 - dx, self.color);
            }
            if bottom_left {
                self.set_pixel_with_color(x0 - dx, y0 + dy, self.color);
                self.set_pixel_with_color(x0 - dy, y0 + dx, self.color);
            }
            if bottom_right {
                self.set_pixel_with_color(x0 + dx, y0 + dy, self.color);
                self.set_pixel_with_color(x0 + dy, y0 + dx, self.color);
            }

            if d < 0 {
                d += 2 * dx + 3;
            } else {
                d += 2 * (dx - dy) + 5;
                dy -= 1;
            }
            dx += 1;
        }
    }
}

impl FontDrawTarget for Canvas {
    fn draw_hline_with_color(&mut self, x: i32, y: i32, length: u32, color: Color) {
        self.draw_hline_color(x, y, length, color);
    }
}

File: firmware/src/canvas.cpp
// 1:1 port of fri3d-runtime/src/canvas.rs. Verified bit-exact against the
// Rust reference via tools/canvas-parity-gen + firmware/tools/canvas-parity.

#include "canvas.h"
#include "font.h"

#include <string.h>

namespace fri3d {

// Tiny adapter that lets Font's draw_str render into a Canvas without making
// Canvas itself inherit from FontDrawTarget (which would require virtual
// dispatch in a hot path, and a full font.h dependency in canvas.h).
class CanvasFontTarget final : public FontDrawTarget {
public:
    explicit CanvasFontTarget(Canvas& c) : m_canvas(c) {}
    void draw_hline_with_color(int32_t x, int32_t y, uint32_t length, Color c) override {
        m_canvas.draw_hline_with_color(x, y, length, c);
    }
private:
    Canvas& m_canvas;
};

Canvas::Canvas() { clear(); }

void Canvas::clear() { memset(m_buffer, 0, sizeof(m_buffer)); }

void Canvas::fill_from(const uint8_t* src, size_t len) {
    if (!src) return;
    const size_t fb_size = sizeof(m_buffer);
    const size_t n = len < fb_size ? len : fb_size;
    memcpy(m_buffer, src, n);
}

void Canvas::set_pixel(int32_t x, int32_t y, Color c) {
    if (x < 0 || y < 0) return;
    if ((uint32_t)x >= SCREEN_WIDTH || (uint32_t)y >= SCREEN_HEIGHT) return;
    const size_t idx = (size_t)y * SCREEN_WIDTH + (size_t)x;
    switch (c) {
        case Color::White: m_buffer[idx] = 0; break;
        case Color::Black: m_buffer[idx] = 1; break;
        case Color::Xor:   m_buffer[idx] ^= 1; break;
    }
}

void Canvas::draw_hline_c(int32_t x, int32_t y, uint32_t length, Color c) {
    for (uint32_t i = 0; i < length; ++i) set_pixel(x + (int32_t)i, y, c);
}

void Canvas::draw_vline_c(int32_t x, int32_t y, uint32_t length, Color c) {
    for (uint32_t i = 0; i < length; ++i) set_pixel(x, y + (int32_t)i, c);
}

void Canvas::draw_dot(int32_t x, int32_t y)                         { set_pixel(x, y, m_color); }
void Canvas::draw_hline(int32_t x, int32_t y, uint32_t length)      { draw_hline_c(x, y, length, m_color); }
void Canvas::draw_vline(int32_t x, int32_t y, uint32_t length)      { draw_vline_c(x, y, length, m_color); }

// Bresenham, matching canvas.rs:73.
void Canvas::draw_line(int32_t x1, int32_t y1, int32_t x2, int32_t y2) {
    int32_t dx = x2 - x1; if (dx < 0) dx = -dx;
    int32_t dy = y2 - y1; if (dy < 0) dy = -dy;
    bool swap_xy = false;

    if (dy > dx) {
        swap_xy = true;
        int32_t t;
        t = x1; x1 = y1; y1 = t;
        t = x2; x2 = y2; y2 = t;
        t = dx; dx = dy; dy = t;
    }
    if (x1 > x2) {
        int32_t t;
        t = x1; x1 = x2; x2 = t;
        t = y1; y1 = y2; y2 = t;
    }

    int32_t err = dx >> 1;
    int32_t ystep = (y2 > y1) ? 1 : -1;
    int32_t y = y1;

    for (int32_t x = x1; x <= x2; ++x) {
        if (!swap_xy) set_pixel(x, y, m_color);
        else          set_pixel(y, x, m_color);
        err -= dy;
        if (err < 0) { y += ystep; err += dx; }
    }
}

void Canvas::draw_frame(int32_t x, int32_t y, uint32_t w, uint32_t h) {
    if (w == 0 || h == 0) return;
    draw_hline_c(x,                   y,                   w, m_color);
    draw_hline_c(x,                   y + (int32_t)h - 1,  w, m_color);
    draw_vline_c(x,                   y,                   h, m_color);
    draw_vline_c(x + (int32_t)w - 1,  y,                   h, m_color);
}

void Canvas::draw_box(int32_t x, int32_t y, uint32_t w, uint32_t h) {
    if (w == 0 || h == 0) return;
    for (uint32_t row = 0; row < h; ++row) {
        draw_hline_c(x, y + (int32_t)row, w, m_color);
    }
}

// ---- Circle / disc with per-quadrant option --------------------------------

void Canvas::draw_circle_section(int32_t x, int32_t y, int32_t x0, int32_t y0, uint8_t option, Color c) {
    if (option & DRAW_UPPER_RIGHT) {
        set_pixel(x0 + x, y0 - y, c);
        set_pixel(x0 + y, y0 - x, c);
    }
    if (option & DRAW_UPPER_LEFT) {
        set_pixel(x0 - x, y0 - y, c);
        set_pixel(x0 - y, y0 - x, c);
    }
    if (option & DRAW_LOWER_RIGHT) {
        set_pixel(x0 + x, y0 + y, c);
        set_pixel(x0 + y, y0 + x, c);
    }
    if (option & DRAW_LOWER_LEFT) {
        set_pixel(x0 - x, y0 + y, c);
        set_pixel(x0 - y, y0 + x, c);
    }
}

void Canvas::draw_circle_with_option(int32_t x0, int32_t y0, int32_t radius, uint8_t option, Color c) {
    if (radius < 0) return;

    int32_t f = 1 - radius;
    int32_t dd_f_x = 1;
    int32_t dd_f_y = -2 * radius;
    int32_t x = 0;
    int32_t y = radius;

    draw_circle_section(x, y, x0, y0, option, c);

    while (x < y) {
        if (f >= 0) { y -= 1; dd_f_y += 2; f += dd_f_y; }
        x += 1; dd_f_x += 2; f += dd_f_x;
        draw_circle_section(x, y, x0, y0, option, c);
    }
}

void Canvas::draw_disc_section(int32_t x, int32_t y, int32_t x0, int32_t y0, uint8_t option, Color c) {
    if (option & DRAW_UPPER_RIGHT) {
        draw_vline_c(x0 + x, y0 - y, (uint32_t)(y + 1), c);
        draw_vline_c(x0 + y, y0 - x, (uint32_t)(x + 1), c);
    }
    if (option & DRAW_UPPER_LEFT) {
        draw_vline_c(x0 - x, y0 - y, (uint32_t)(y + 1), c);
        draw_vline_c(x0 - y, y0 - x, (uint32_t)(x + 1), c);
    }
    if (option & DRAW_LOWER_RIGHT) {
        draw_vline_c(x0 + x, y0, (uint32_t)(y + 1), c);
        draw_vline_c(x0 + y, y0, (uint32_t)(x + 1), c);
    }
    if (option & DRAW_LOWER_LEFT) {
        draw_vline_c(x0 - x, y0, (uint32_t)(y + 1), c);
        draw_vline_c(x0 - y, y0, (uint32_t)(x + 1), c);
    }
}

void Canvas::draw_disc_with_option(int32_t x0, int32_t y0, int32_t radius, uint8_t option, Color c) {
    if (radius < 0) return;

    int32_t f = 1 - radius;
    int32_t dd_f_x = 1;
    int32_t dd_f_y = -2 * radius;
    int32_t x = 0;
    int32_t y = radius;

    draw_disc_section(x, y, x0, y0, option, c);

    while (x < y) {
        if (f >= 0) { y -= 1; dd_f_y += 2; f += dd_f_y; }
        x += 1; dd_f_x += 2; f += dd_f_x;
        draw_disc_section(x, y, x0, y0, option, c);
    }
}

// ---- XOR circle / disc -----------------------------------------------------
// Bresenham visits several pixels twice (at octant boundaries). For solid
// colours that's harmless; for XOR it cancels the pixel. These paths visit
// each pixel exactly once.

void Canvas::draw_xor_circle(int32_t x0, int32_t y0, uint32_t radius) {
    if (radius == 0) { set_pixel(x0, y0, Color::Xor); return; }

    int32_t x = 0;
    int32_t y = (int32_t)radius;
    int32_t d = 1 - y;

    while (x <= y) {
        if (x == y) {
            set_pixel(x0 + x, y0 + y, Color::Xor);
            set_pixel(x0 - x, y0 + y, Color::Xor);
            set_pixel(x0 + x, y0 - y, Color::Xor);
            set_pixel(x0 - x, y0 - y, Color::Xor);
        } else if (x == 0) {
            set_pixel(x0,     y0 + y, Color::Xor);
            set_pixel(x0,     y0 - y, Color::Xor);
            set_pixel(x0 + y, y0,     Color::Xor);
            set_pixel(x0 - y, y0,     Color::Xor);
        } else {
            set_pixel(x0 + x, y0 + y, Color::Xor);
            set_pixel(x0 - x, y0 + y, Color::Xor);
            set_pixel(x0 + x, y0 - y, Color::Xor);
            set_pixel(x0 - x, y0 - y, Color::Xor);
            set_pixel(x0 + y, y0 + x, Color::Xor);
            set_pixel(x0 - y, y0 + x, Color::Xor);
            set_pixel(x0 + y, y0 - x, Color::Xor);
            set_pixel(x0 - y, y0 - x, Color::Xor);
        }

        if (d < 0) {
            d += 2 * x + 3;
        } else {
            d += 2 * (x - y) + 5;
            y -= 1;
        }
        x += 1;
    }
}

void Canvas::draw_xor_disc(int32_t x0, int32_t y0, uint32_t radius) {
    if (radius == 0) { set_pixel(x0, y0, Color::Xor); return; }

    const int32_t r = (int32_t)radius;
    const int32_t r_sq = r * r;

    for (int32_t dy = -r; dy <= r; ++dy) {
        const int32_t y_sq = dy * dy;
        if (y_sq > r_sq) continue;

        int32_t x_extent = 0;
        while ((x_extent + 1) * (x_extent + 1) <= r_sq - y_sq) {
            ++x_extent;
        }

        const int32_t line_y = y0 + dy;
        const int32_t line_x = x0 - x_extent;
        const uint32_t line_w = (uint32_t)(2 * x_extent + 1);

        draw_hline_c(line_x, line_y, line_w, Color::Xor);
    }
}

void Canvas::draw_circle(int32_t x0, int32_t y0, uint32_t radius) {
    if (m_color == Color::Xor) { draw_xor_circle(x0, y0, radius); return; }
    draw_circle_with_option(x0, y0, (int32_t)radius, DRAW_ALL, m_color);
}

void Canvas::draw_disc(int32_t x0, int32_t y0, uint32_t radius) {
    if (m_color == Color::Xor) { draw_xor_disc(x0, y0, radius); return; }
    draw_disc_with_option(x0, y0, (int32_t)radius, DRAW_ALL, m_color);
}

// ---- Rounded frame / box ---------------------------------------------------

void Canvas::draw_rframe(int32_t x, int32_t y, uint32_t width, uint32_t height, uint32_t radius) {
    if (width == 0 || height == 0) return;
    int32_t r = (int32_t)radius;
    const int32_t w = (int32_t)width;
    const int32_t h = (int32_t)height;
    if (r > w / 2) r = w / 2;
    if (r > h / 2) r = h / 2;
    if (r <= 0) { draw_frame(x, y, width, height); return; }

    int32_t xl = x + r;
    int32_t yu = y + r;
    const int32_t xr = x + w - r - 1;
    const int32_t yl = y + h - r - 1;

    draw_circle_with_option(xl, yu, r, DRAW_UPPER_LEFT,  m_color);
    draw_circle_with_option(xr, yu, r, DRAW_UPPER_RIGHT, m_color);
    draw_circle_with_option(xl, yl, r, DRAW_LOWER_LEFT,  m_color);
    draw_circle_with_option(xr, yl, r, DRAW_LOWER_RIGHT, m_color);

    int32_t ww = w - r - r;
    int32_t hh = h - r - r;

    xl += 1;
    yu += 1;

    if (ww >= 3) {
        ww -= 2;
        const int32_t edge_y = y + h - 1;
        draw_hline_c(xl, y,      (uint32_t)ww, m_color);
        draw_hline_c(xl, edge_y, (uint32_t)ww, m_color);
    }
    if (hh >= 3) {
        hh -= 2;
        const int32_t edge_x = x + w - 1;
        draw_vline_c(x,       yu, (uint32_t)hh, m_color);
        draw_vline_c(edge_x,  yu, (uint32_t)hh, m_color);
    }
}

void Canvas::draw_rbox(int32_t x, int32_t y, uint32_t width, uint32_t height, uint32_t radius) {
    if (width == 0 || height == 0) return;
    int32_t r = (int32_t)radius;
    const int32_t w = (int32_t)width;
    const int32_t h = (int32_t)height;
    if (r > w / 2) r = w / 2;
    if (r > h / 2) r = h / 2;
    if (r <= 0) { draw_box(x, y, width, height); return; }

    int32_t xl = x + r;
    int32_t yu = y + r;
    const int32_t xr = x + w - r - 1;
    const int32_t yl = y + h - r - 1;

    draw_disc_with_option(xl, yu, r, DRAW_UPPER_LEFT,  m_color);
    draw_disc_with_option(xr, yu, r, DRAW_UPPER_RIGHT, m_color);
    draw_disc_with_option(xl, yl, r, DRAW_LOWER_LEFT,  m_color);
    draw_disc_with_option(xr, yl, r, DRAW_LOWER_RIGHT, m_color);

    int32_t ww = w - r - r;
    int32_t hh = h - r - r;

    xl += 1;
    yu += 1;

    if (ww >= 3) {
        ww -= 2;
        draw_box(xl, y,  (uint32_t)ww, (uint32_t)(r + 1));
        draw_box(xl, yl, (uint32_t)ww, (uint32_t)(r + 1));
    }
    if (hh >= 3) {
        hh -= 2;
        draw_box(x, yu, (uint32_t)w, (uint32_t)hh);
    }
}

// ---- Text ------------------------------------------------------------------

void Canvas::draw_str(int32_t x, int32_t y, const char* text) {
    CanvasFontTarget target(*this);
    Font::from_id(m_font).draw_str(target, x, y, text, m_color);
}

uint32_t Canvas::string_width(const char* text) const {
    return Font::from_id(m_font).string_width(text);
}

} // namespace fri3d
